substrate/client/chain-spec/src/lib.rs

```diff
@@ -16,19 +16,228 @@
+//! intermediate internal helper that facilitates this interaction,
+//! [`GenesisConfigBuilderRuntimeCaller`], which serves as a straightforward wrapper for
+//! [`sc_executor::WasmExecutor`].
+//!
+//! In case of `raw` genesis state the node does not interact with the runtime regarding the
+//! computation of initial state.
+//!
+//! The plain and `raw` chain specification JSON blobs can be found in
+//! [JSON examples](#json-chain-specification-example) section.
+//!
+//! # Optional code mapping
+//! Optional map of `block_number` to `wasm_code`.
+//!
+//! The given `wasm_code` will be used to substitute the on-chain wasm code starting with the
+//! given block number until the `spec_version` on-chain changes. The given `wasm_code` should
+//! be as close as possible to the on-chain wasm code. A substitute should be used to fix a bug
+//! that can not be fixed with a runtime upgrade, if for example the runtime is constantly
+//! panicking. Introducing new runtime apis isn't supported, because the node
+//! will read the runtime version from the on-chain wasm code. Use this functionality only when
+//! there is no other way around it and only patch the problematic bug, the rest should be done
+//! with a on-chain runtime upgrade.
+//!
+//! # Building a chain specification.
+//! The [`ChainSpecBuilder`] shall be used to create an instance of chain specification. Its API
+//! allows to configure all fields of the chain spec. To generate a JSON representation of chain
+//! specification, use [`ChainSpec::as_json`].
+//!
+//! The sample code to generate a chain spec is as follows:
+#![doc = docify::embed!("src/chain_spec.rs", build_chain_spec_with_patch_works)]
+//! # JSON chain specification example
+//! The following are the plain and `raw` ersions of the chain specification JSON files, resulting
+//! from executing of the above [example](#building-a-chain-specification):
+//! ```ignore
+#![doc = include_str!("../res/substrate_test_runtime_from_patch.json")]
+//! ```
+//! ```ignore
+#![doc = include_str!("../res/substrate_test_runtime_from_patch_raw.json")]
+//! ```
+//! The [`ChainSpec`] trait represents the API to access values defined in JSON chain specification.
 //!
-//! Basic chain spec type containing all required parameters is
-//! [`GenericChainSpec`]. It can be extended with
-//! additional options that contain configuration specific to your chain.
-//! Usually the extension is going to be an amalgamate of types exposed
-//! by Substrate core modules. To allow the core modules to retrieve
-//! their configuration from your extension you should use `ChainSpecExtension`
-//! macro exposed by this crate.
+//! # Custom chain spec extensions
+//! Basic chain spec type containing all required parameters is [`GenericChainSpec`]. It can be
+//! extended with additional options that contain configuration specific to your chain. Usually the
+//! extension is going to be an amalgamate of types exposed by Substrate core modules.
```

Comment:
```suggestion
//! The basic chain spec type containing all required parameters is [`GenericChainSpec`]. It can be
//! extended with additional options containing configuration specific to your chain. Usually, the
//! extension will be a combination of types exposed by Substrate core modules.
```